File: core/full_pipeline.py
"""
WebWeaveX V7/V8 - FINAL PURE + INTELLIGENCE PIPELINE

Pipeline:
INPUT → TOKENIZATION → NODES → RELATIONSHIPS → GRAPH → SYSTEM → EXECUTION → INTELLIGENCE → OUTPUT

STRICT RULES:
- Deterministic
- No heuristics
- No type fields
- No keyword logic
"""

from typing import Dict, Any
from core.version import ENGINE_VERSION

try:
    from core.intelligence.intelligence_engine import run_intelligence
    from core.crypto.kaalka_wrapper import graph_fingerprint
    KAALKA_AVAILABLE = True
except ImportError:
    KAALKA_AVAILABLE = False

PIPELINE_VERSION = ENGINE_VERSION


def run_pipeline(user_input: str, mode: str = "compiler") -> dict:
    """Main pipeline."""

    tokens = _tokenize(user_input)
    nodes = _generate_nodes(tokens)
    relationships = _generate_relationships(tokens)

    system = _derive_system(nodes, relationships)
    execution_graph = _build_execution_graph(nodes, relationships)
    execution_order = _derive_execution_order(nodes)
    spec = _build_spec(nodes, relationships)

    # SINGLE SOURCE GRAPH (CRITICAL FIX)
    # Intelligence ALWAYS independent of Kaalka
    intelligence = {}
    
    try:
        from core.intelligence.intelligence_engine import run_intelligence
        intelligence = run_intelligence(execution_graph)
    except ImportError:
        pass
    
    # Kaalka ONLY for fingerprint
    fingerprint = ""
    
    if KAALKA_AVAILABLE:
        fp_bytes = graph_fingerprint(execution_graph)
        fingerprint = fp_bytes.hex()

    output = {
        "structured_data": {
            "system": system,
            "execution_graph": execution_graph,
            "execution_order": execution_order,
            "spec": spec,
            "intelligence": intelligence,
            "fingerprint": fingerprint
        },
        "confidence": 1.0,
        "source": "compiler",
        "version": PIPELINE_VERSION
    }
    
    return output


def _tokenize(user_input: str) -> list:
    if not user_input:
        return []

    tokens = user_input.lower()
    tokens = tokens.replace(",", " ").replace("-", " ").replace("_", " ")
    tokens = tokens.replace(".", " ").replace("(", " ").replace(")", " ")
    tokens = tokens.replace("/", " ").split()

    return [t for t in tokens if t]


def _generate_nodes(tokens: list) -> list:
    unique_tokens = sorted(set(tokens))
    nodes = [{"id": t} for t in unique_tokens]
    return sorted(nodes, key=lambda x: x["id"])


def _generate_relationships(tokens: list) -> list:
    if not tokens:
        return []

    WINDOW = 3
    relationships = []

    # SYMMETRIC WINDOW (no positional bias)
    for i in range(len(tokens)):
        for j in range(max(0, i - WINDOW), min(len(tokens), i + WINDOW + 1)):
            if i != j:
                relationships.append({
                    "from": tokens[i],
                    "to": tokens[j]
                })

    # DEDUPLICATE
    seen = set()
    unique = []

    for r in relationships:
        key = (r["from"], r["to"])
        if key not in seen:
            seen.add(key)
            unique.append(r)

    return sorted(unique, key=lambda x: (x["from"], x["to"]))


def _derive_system(nodes: list, relationships: list) -> dict:
    components = sorted(
        [{"name": n.get("id", "")} for n in nodes],
        key=lambda x: x["name"]
    )

    return {
        "system_type": "",
        "architecture": "",
        "components": components,
        "relationships": relationships
    }


def _build_execution_graph(nodes: list, relationships: list) -> dict:
    edges = [{"from": r.get("from", ""), "to": r.get("to", "")} for r in relationships]

    return {
        "nodes": nodes,
        "edges": sorted(edges, key=lambda x: (x["from"], x["to"]))
    }


def _derive_execution_order(nodes: list) -> list:
    return sorted([n.get("id", "") for n in nodes])


def _build_spec(nodes: list, relationships: list) -> dict:
    return {
        "node_count": len(nodes),
        "edge_count": len(relationships)
    }

File: src/fullPipeline.ts
/**
 * Converted from Python: core/full_pipeline.py
 * @generated — WebWeaveX python→javascript library port
 */

import * as py from "./runtime/pyCompat.js";
import { ENGINE_VERSION } from "./version.js";
import { runIntelligence } from "./intelligence/intelligenceEngine.js";
import { graphFingerprint } from "./crypto/kaalkaWrapper.js";

var KAALKA_AVAILABLE: any = true;
export let PIPELINE_VERSION: any = ENGINE_VERSION;
export function runPipeline(user_input: any, mode: any = "compiler"): any {
  var tokens: any = _tokenize(user_input);
  var nodes: any = _generateNodes(tokens);
  var relationships: any = _generateRelationships(tokens);
  var system: any = _deriveSystem(nodes, relationships);
  var execution_graph: any = _buildExecutionGraph(nodes, relationships);
  var execution_order: any = _deriveExecutionOrder(nodes);
  var spec: any = _buildSpec(nodes, relationships);
  var intelligence: Record<string, any> = {};
  intelligence = runIntelligence(execution_graph);
  var fingerprint: any = "";
  if (py.truthy(KAALKA_AVAILABLE)) {
    var fp_bytes: any = graphFingerprint(execution_graph);
    fingerprint = fp_bytes.hex();
  }
  var output: any = {"structured_data": {"system": system, "execution_graph": execution_graph, "execution_order": execution_order, "spec": spec, "intelligence": intelligence, "fingerprint": fingerprint}, "confidence": py.F(1.0), "source": "compiler", "version": PIPELINE_VERSION};
  return output;
}
export function _tokenize(user_input: any): any {
  if (!py.truthy(user_input)) {
    return [];
  }
  var tokens: any = String(user_input).toLowerCase();
  tokens = py.replace(py.replace(py.replace(tokens, ",", " "), "-", " "), "_", " ");
  tokens = py.replace(py.replace(py.replace(tokens, ".", " "), "(", " "), ")", " ");
  tokens = py.split(py.replace(tokens, "/", " "));
  return py.iter(tokens).filter((t: any) => py.truthy(t)).map((t: any) => t);
}
export function _generateNodes(tokens: any): any {
  var unique_tokens: any = py.sorted(py.toSet(tokens));
  var nodes: any = py.iter(unique_tokens).map((t: any) => ({"id": t}));
  return py.sorted(nodes, {key: ((x: any) => py.at(x, "id")) as (item: any) => any});
}
export function _generateRelationships(tokens: any): any {
  if (!py.truthy(tokens)) {
    return [];
  }
  var WINDOW: any = 3;
  var relationships: any[] = [];
  var i: any;
  for (i = 0; i < py.len(tokens); i++) {
    var j: any;
    for (j = py.max([0, py.sub(i, WINDOW)]); j < py.min([py.len(tokens), py.add(py.add(i, WINDOW), 1)]); j++) {
      if (!py.eq(i, j)) {
        py.listAppend(relationships, {"from": py.at(tokens, i), "to": py.at(tokens, j)});
      }
    }
  }
  var seen: Set<any> = new Set();
  var unique: any[] = [];
  var r: any;
  for (r of py.iter(relationships)) {
    var key: any = [py.at(r, "from"), py.at(r, "to")];
    if (!py.contains(seen, key)) {
      py.setAdd(seen, key);
      py.listAppend(unique, r);
    }
  }
  return py.sorted(unique, {key: ((x: any) => [py.at(x, "from"), py.at(x, "to")]) as (item: any) => any});
}
export function _deriveSystem(nodes: any, relationships: any): any {
  var components: any = py.sorted(py.iter(nodes).map((n: any) => ({"name": py.get(n, "id", "")})), {key: ((x: any) => py.at(x, "name")) as (item: any) => any});
  return {"system_type": "", "architecture": "", "components": components, "relationships": relationships};
}
export function _buildExecutionGraph(nodes: any, relationships: any): any {
  var edges: any = py.iter(relationships).map((r: any) => ({"from": py.get(r, "from", ""), "to": py.get(r, "to", "")}));
  return {"nodes": nodes, "edges": py.sorted(edges, {key: ((x: any) => [py.at(x, "from"), py.at(x, "to")]) as (item: any) => any})};
}
export function _deriveExecutionOrder(nodes: any): any {
  return py.sorted(py.iter(nodes).map((n: any) => py.get(n, "id", "")));
}
export function _buildSpec(nodes: any, relationships: any): any {
  return {"node_count": py.len(nodes), "edge_count": py.len(relationships)};
}
export { ENGINE_VERSION, graphFingerprint, runIntelligence };
